server/src/main/java/org/opensearch/indices/IntKeyLookupStore.java

```diff
@@ -0,0 +1,164 @@
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+/*
+ * Modifications Copyright OpenSearch Contributors. See
+ * GitHub history for details.
+ */
+
+package org.opensearch.indices;
+
+/**
+ * An interface for objects that hold an in-memory record of hashes of keys in the disk cache.
+ * These objects have some internal data structure which stores some transformation of added
+ * int values. The internal representations may have collisions. Example transformations include a modulo
+ * or -abs(value), or some combination.
+ */
+public interface IntKeyLookupStore {
+    /**
+     * Transforms the input value into the internal representation for this keystore
+     * and adds it to the internal data structure.
+     * @param value The value to add.
+     * @return true if the value was added, false if it wasn't added because of a
+     * collision or if it was already present.
+     */
+    boolean add(int value) throws Exception;
+
+    /**
+     * Checks if the transformation of the value is in the keystore.
+     * @param value The value to check.
+     * @return true if the value was found, false otherwise. Due to collisions, false positives are
+     * possible, but there should be no false negatives unless forceRemove() is called.
+     */
+    boolean contains(int value) throws Exception;
+
+    /**
+     * Returns the transformed version of the input value, that would be used to stored it in the keystore.
+     * This transformation should be always be the same for a given instance.
+     * @param value The value to transform.
+     * @return The transformed value.
+     */
+    int getInternalRepresentation(int value);
+
+    /**
+     * Attempts to safely remove a value from the internal structure, maintaining the property that contains(value)
+     * will never return a false negative. If removing would lead to a false negative, the value won't be removed.
+     * Classes may not implement safe removal.
+     * @param value The value to attempt to remove.
+     * @return true if the value was removed, false if it wasn't.
+     */
+    boolean remove(int value) throws Exception;
+
+    /**
+     * Check if the implementing class supports safe removals. If it doesn't, remove() will always return false.
+     * @return true if the class supports safe removals, false if it doesn't.
+     */
+    boolean supportsRemoval();
+
+    /**
+     * Remove the transformed version of this value from the store. Calling this function may cause
+     * contains() to return false negatives for future values.
+     * @param value The value to forcibly remove.
+     */
+    void forceRemove(int value) throws Exception;
+
+    /**
+     * Check if the object currently guarantees having no false negatives when running contains().
+     * @return false if there will not be false negatives, true if there could be false negatives.
+     */
+    boolean canHaveFalseNegatives();
```

Comment:
Again, I think we should disallow false negatives at all cost? As it should be a basic requirement, and can be costly depending on the use case(like computing query again though it was present in cache)